[String] Use UTF-8 as native encoding

milseman opened the pull request:
<!-- What's in this pull request? -->

This is just a prototype for an alternative String design that can accommodate UTF-8 efficiently.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

airspeedswift committed:
[stdlib] Add consuming/owned annotations to Collection implementations (#19360)

* Add consuming/owned annotations to Collection implementations

* Update SILOptimizer tests

* Fix access_marker_verify test

* XFAIL reconstruct_type_from_mangled_name

milseman commented:
@swift-ci please smoke benchmark

Commit:
String comparison iterator for UTF8 strings

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
Those results don't have my changes... Trying again.

@swift-ci please smoke benchmark

milseman committed:
[String] Cleanup normalization code.

Clean up some of the code surrounding the normalized code unit
iterator.

milseman commented:
@swift-ci please smoke benchmark

milseman reviewed (commented):
cc @lancep

shahmishal commented:
@swift-ci please smoke benchmark

airspeedswift committed:
Give StringProtocol.SubSequence a default of Substring to supress warning

milseman committed:
[tests] Fix, cleanup, and clarify some SIL character tests

weissi committed:
integrating utf8 validation

spevans committed:
SR-8876: Always build libicu on Linux

- Uses version 61.1 from ICU Github unicode-org/icu repository.

- Updates mixin_linux_installation to add libicu option.

- Use -j when building libicu.

- When buiding ICU, use --with-library-suffix=swift
  This suffixes the ICU symbols with _swift.

  The libaries are now named libicuucswift, libicui18nswift
  and libicudataswift.

- Add the contents of uconfig.h.prepend into uconfig.h. This avoids
  passing the renaming CFLAGS to swift and swift-corelibs-foundation.

Also resolves:

SR-5618: libicu compilation should happen in parallel respecting -j.

lorentey commented on stdlib/public/core/StringStorage.swift:
```diff
@@ -12,308 +12,474 @@
+  private var unusedStorage: UnsafeMutableBufferPointer<UInt8> {
+    @inline(__always) get {
+      return UnsafeMutableBufferPointer(
+        start: mutableEnd, count: capacity)
```
I think this should use `unusedCapacity` as the count.

milseman replied:
Doh, good catch.

lorentey commented on stdlib/public/core/StringIndex.swift:
```diff
@@ -9,133 +9,266 @@
   internal var characterStride: Int? {
-    guard _graphemeStrideCache > 0 else { return nil }
-    return Int(truncatingIfNeeded: _graphemeStrideCache)
+    let value = (_rawBits & 0x00FC_0000_0000_0000) &>> 50
```
This doesn't match the position of the grapheme cache in the layout description above!

milseman replied:
Huh, that might be why grapheme breaking is 2x slower in some ad-hoc tests I did. It's almost as if it's measuring the distance twice because nothing's in the cache... 😅

Good catch!

milseman commented on stdlib/public/core/StringGutsRangeReplaceable.swift:
```diff
@@ -0,0 +1,270 @@
+
+    if isUniqueNative {
+      _object.nativeStorage.remove(from: lowerOffset, to: upperOffset)
+      return
```
Yeah, I have a fix that I thought I pushed. It might be just to my comparison branch, though

milseman replied:
This is why I'm re-thinking making the operations be on _StringStorage...

Commit:
Add benchmarks and tests for the normalized iterator (#32)

Add benchmarks and tests for the normalized iterator

milseman committed:
[String] Performance improvements to comparison

Also, disable normalization benchmarks and other changes until we
merge, so we can compare with master 1-to-1.

spevans committed:
update_checkout.py: Add support for tags and optional repositories.

- Update --reset-to-remote to support tags.

- Add optional platform entry to a repository. This allows specifying
  which platforms a repo should be checked out on.

Commit:
fix the normalization unit tests

milseman committed:
[String] More comparison speedups and cleanup

milseman committed:
[test] Fix test: add in explicit mkdir

milseman committed:
[String] Fix overlays to request index interchange for native strings

milseman committed:
[String] Refactor helper code into UnicodeHelpers.swift.

Clean up some of the index assumptions, stick index-aware methods on
_StringGuts, and otherwise migrate code over to UnicodeHelpers.swift.

milseman committed:
[String] Use small string at string creation

milseman committed:
[String] In-place append and other RRC improvements

milseman committed:
[String] Bounds checking and Index cleanup

milseman committed:
[String] Emit literals as UTF-8 rather than UTF-16

milseman commented:
@swift-ci please smoke benchmark

milseman commented:
@swift-ci please smoke benchmark

milseman committed:
[test] Replace hard-code SIL test with Swift-based one

Replaces tests for constant-folding string concatenation with
Swift-based ones, which are easier to maintain going forwards.

weissi committed:
use legacy replacement ranges to fix tests

this is just to be compatible with the current transcoders and can be
undone when the transcoders (based on ForwardParser/ReverseParser) are
improved too.

milseman committed:
[test] Test the unique-native String RRC optimization path

milseman committed:
[String] Implement in-place generic RRC

milseman commented:
@swift-ci please test linux platform

milseman commented:
@swift-ci please test linux platform

lorentey commented:
An LLDB data formatter update corresponding to these changes can be found at https://github.com/apple/swift-lldb/pull/998.

milseman committed:
[test] Test the breadcrumbing String<->Cocoa interface

compnerd committed:
build: switch swift-corelibs-foundation to CMake

swift-corelibs-foundation has had a CMake based build system for a while
now.  Switch the official builds over to it.

milseman committed:
[String] 5X Faster getCharacters implementation

Rather than bounce through the UTF-16 view, implement custom
transcoding for getCharacters. This speeds it up by around 5X. Adds
tests.

milseman committed:
[String] Cleanups, comments, documentation

After rebasing on master and incorporating more 32-bit support,
perform a bunch of cleanup, documentation updates, comments, move code
back to String declaration, etc.

lorentey committed:
[String] 32-bit platform support

Add support for 32-bit platforms for UTF-8 backed String.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke test

milseman commented:
@swift-ci please test source compatibility

lorentey commented on stdlib/public/core/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,139 @@
+        Builtin.addressof(&_object.sharedStorage._breadcrumbs))
+    }
+
+    if _slowPath(mutPtr.pointee == nil) {
```
Shouldn't this use an atomic load?

milseman replied:
I don't think it has to be. `populateBreadcrumbs` constructs a new instance and atomic compare-and-exchanges it. Worst case there will be redundant work that will be thrown away at the end. That's rare and probably a better price to pay than an atomic operation on every single read of the breadcrumbs.

CC @atrick, thoughts?

milseman replied:
I don't think it has to be. populateBreadcrumbs constructs a new instance and atomic compare-and-exchanges it. Worst case there will be redundant work that will be thrown away at the end. That's rare and probably a better price to pay than an atomic operation on every single read of the breadcrumbs.

CC @atrick, thoughts?

atrick replied:
Given that the breadcrumbs are immutable data that are published once and
remain valid, and that breadcrumb data is published via a pointer
value that is atomically written after initializing the data....

Then we definitely don't want to do any hardware atomic operations on
the read of the breadcrumbs. In theory, this load does need to be a
"memory_order_consume", but I think a comment would suffice. (It would
be ridiculous if LLVM stopped assuming that all loads had "consume"
semantics).

To prove safety, we just need these conditions:

- A happens-before relationship exists between the store of the
  breadcrumb values and the store of the publishing pointer. Since
  _stdlib_atomicInitializeARCRef is the only way to write the
  _StringBreadcrumbs reference inside the storage, that's
  covered. Minimally, an atomic store-release is
  required. cmpxchg_seqcst_seqcst is actually overkill.

- The only way to read a breadcrumb is via a dependent load on the
  published _StringBreadcrumbs reference (consume ordering).

- Nothing stores to the breadcrumb data again after it is
  published. You could potentially allow atomic relaxed consistency
  stores to the breadcumb later if the order the changes are seen didn't
  matter. Then the reads would also need to be atomic relaxed
  consistency, which in practice should not cost anything but would
  prevent word-tearing. In your case that wouldn't make sense anyway.

It would be a good idea to comment these assumptions, in some more concise form.

lorentey replied:
Perhaps we should make an AtomicInitializableRef construct around an UMP to capture this usecase, which occurs a few times elsewhere in the stdlib. Array/Set/Dictionary do something very similar to implement deferred bridging, except they use atomic loads — they too would benefit from a smarter implementation.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test linux platform

milseman commented:
@swift-ci please smoke test linux platform

milseman committed:
[test] Adjust test: we do more constant folding than before!

milseman commented:
@swift-ci please test

atrick commented on stdlib/public/core/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,139 @@
+  var utf16Length: Int
+
+  // TODO: does this need to be a pair?.... Can we be smaller than Int?
+  var crumbs: [String.Index]
```
Why is `crumbs` an Array instead of tail allocated, fixed sized, trivial storage. Don't we want the _StrigBreadcrumbs implicit `deinit` to be a nop? (I realize this is non-ABI impacting, hence not important. Just bringing it up while I noticed).

milseman replied:
Sure, that would work as an optimization. This whole class is an optional second tail allocation on String itself, but this would speed up access and destruction.

milseman commented:
@swift-ci please test linux platform

milseman committed:
[string] Less inlining for code size.

Less inlining for hashing and comparison. Saves code size on very
frequent String comparison in exchange for costing us in some of our
ridiculuous micro-benchmarks.

Also adds in more _effects for better codegen

milseman commented:
@swift-ci please benchmark

milseman commented:
@swift-ci please test compiler performance

compnerd committed:
add_swift_host_library: remove FILE_DEPENDS

compnerd committed:
add_swift_host_library: remove C_COMPILE_FLAGS

compnerd committed:
add_swift_host_library: rename LINK_LIBRARIES to LINK_LIBS

compnerd committed:
add_swift_host_library: rename LLVM_COMPONENT_DEPENDS to LINK_COMPONENTS

compnerd committed:
add_swift_host_library: remove LINK_INTERFACE_LIBRARIES

compnerd committed:
add_swift_host_library: remove LINK_FLAGS

compnerd committed:
add_swift_host_library: make a trivial wrapper around add_llvm_library

compnerd committed:
add_swift_host_library: special handling for GYB sources

milseman committed:
[String] Refactorings and cleanup

* Refactor out RRC implementation into dedicated file.

* Change our `_invariantCheck` pattern to generate efficient code in
  asserts builds and make the optimizer job's easier.

* Drop a few Bidi shims we no longer need.

* Restore View decls to String, workaround no longer needed

* Cleaner unicode helper facilities

milseman committed:
[String] Introduce StringBreadcrumbs

Breadcrumbs provide us amortized O(1) access to the UTF-16 view, which
is vital for efficient Cocoa interoperability.

milseman committed:
[String] Remove dead code and decls

milseman committed:
[String] Constant-fold small strings from literals.

Tweak and adjust code so that the SIL optimizer can constant-fold
small strings from literals. Also some cleanup.

milseman committed:
[String] Perform small string append in-register

milseman committed:
[String] NFC iterator fast-paths

Refactor and rename _StringGutsSlice, apply NFC-aware fast paths to a
new buffered iterator.

Also, fix bug in _typeName which used to assume ASCIIness and better
SIL optimizations on StringObject.

milseman committed:
[String] Grapheme breaking fast-paths

Add in our scalar-based fast-paths for UTF-8 and foreign strings, and
update the grapheme cache.

milseman committed:
[String] Define performance flags and plumb them throughout

milseman committed:
[String] Inlining, NFC fast paths, and more.

Add inlinability annotations to restore performance parity with 4.2 String.

Take advantage of known NFC as a fast-path for comparison, and
overhaul comparison dispatch.

RRC improvements and optmizations.

milseman committed:
[test] Adjust String tests for UTF-8 representation.

Adjust tests for the UTF-8 representation, in preparation for 32-bit
support. Includes UTF-8 literal update.

milseman committed:
[test] Update test to reflect DCE

milseman committed:
[test] Dump in new API and ABI reference point

Put in the new API/ABI from UTF-8 String.

milseman commented:
@swift-ci please test linux platform

lorentey committed:
[stdlib] Ensure that reserved capacity survives CoW copies (#46)

This shouldn’t really be necessary, but it makes sense to wait until we can reclaim memory like this consistently across the entire stdlib.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci test source compiler performance

milseman commented:
@swift-ci please test source compatibility

milseman commented:
@swift-ci Please Test Source Compatibility Release

milseman commented:
@swift-ci Please Test Source Compatibility Release

milseman commented:
@swift-ci Please Test Source Compatibility Release

milseman committed:
[String] More efficient breadcrumb-scanning code.

Rather than rely on the UTF16View, scan between breadcrumbs by hand
for a decent 20% speedup. This code will also make it more obvious how
to slot in a vectorized solution later.

milseman commented:
@swift-ci Please Test Source Compatibility Release

moiseev committed:
[overlay] Build SceneKit in Swift 4 mode

Addresses: <rdar://problem/43534146>

vlm commented on stdlib/public/core/UnicodeHelpers.swift:
```diff
@@ -0,0 +1,377 @@
+        len += 1
+      }
+      _sanityCheck(len <= 4)
+      return len
```
I believe it is rather important to explicitly check for `len <= 4` in all cases, rather than this being a compile-dependent check. Some dependent code using `fastUTF8ScalarLength` does `Builtin.unreachable()` if length is not one of `{1,2,3,4}`. Therefore the program might crash on some invalid input if the length is not properly bound to `[1..4]` range over here.

milseman replied:
This is only for native strings, which we validate on creation

vlm commented on stdlib/public/core/UnicodeHelpers.swift:
```diff
@@ -0,0 +1,377 @@
+  case 3: return (_decodeUTF8(cu0, utf8[i &+ 1], utf8[i &+ 2]), len)
+  case 4:
+    return (_decodeUTF8(cu0, utf8[i &+ 1], utf8[i &+ 2], utf8[i &+ 3]), len)
+  default: Builtin.unreachable()
```
See comment https://github.com/apple/swift/pull/18680/files#r229134344

milseman committed:
Make corelibs-foundation build

Expose the old SPIs again so corelibs-foundation can build. We'll want
to wean them off of these and establish proper APIs soon.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please build toolchain

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test

milseman commented:
Please test with following PR:
https://github.com/apple/swift-lldb/pull/998

@swift-ci please test

shahmishal commented:
Please test with following PR:
https://github.com/apple/swift-lldb/pull/998

@swift-ci please test

atrick committed:
Fix static initializer analysis, used in LetPropertiesOpt.

Don't exponentially recurse through the use-def graph.

Fixes <rdar://problem/45691574> [SR-9146]
SILCloner crash in IBAnimatable: LetPropertiesOpt duplicates the property initializer sequence.

milseman commented:
Please test with following PR:
apple/swift-lldb#998

@swift-ci please test Linux Platform

xedin committed:
[CSBindings] Detect situations when type variable bindings could be incomplete

When `bind param` constraint is associated with a given type
variable a set of bindings collected for it is potentially
incomplete, because binding gathering doesn't look through related
type variables. Which means that such type variable has to be
de-prioritized until `bind param` constraint is resolved
or there is just nothing else to try, otherwise there is a
risk that solver would skip some of the valid solutions.

Resolves: rdar://problem/45659733

rjmccall committed:
Synthesize modify accessors for resilient global variables.

rdar://32936947

lorentey committed:
[stdlib] Don’t force-inline Character.== and .<

The code to compare character values can be huge; it doesn’t seem a good idea to forcibly inline it. The @inline(__always) here can make specialized Set<Character> operations balloon 10x their current size with unrelated modifications in Set.

The code size bloat doesn’t always happen; I *guess* that _fastPaths inside String.== cause unintuitive inlining decisions in callers.

milseman commented:
Please test with following PR:
apple/swift-lldb#998

@swift-ci please test Linux Platform

Commit:
[Exclusivity] Add new dominating access checks in loop pre-headers

General case:
<loop preheader>
A = ref_element_addr
<loop>
begin_access A [dynamic] [no_nested_conflict]

Adding an empty begin_access A in the preheader would allow us to turn the loop's access to [static]

nkcsgexi committed:
build-script: infer swiftpm executables if swiftpm isn't built.

eeckstein committed:
benchmarks: combine everything which is needed into run_smoke_bench

Now, run_smoke_bench runs the benchmarks, compares performance and code size and reports the results - on stdout and as a markdown file.
No need to run bench_code_size.py and compare_perf_tests.py separately.

This has two benefits:
- It's much easier to run it locally
- It's now more transparent what's happening in '@swiftci benchmark', because now all the logic is in run_smoke_bench rather than in the not visible script on the CI bot.

I also remove the branch-arguments from ReportFormatter in ompare_perf_tests.py. They were not used anyway.

For a smooth rollout in CI, I created a new script rather than changing the existing one. Once everything is setup in CI, I'll delete the old run_smoke_test.py and bench_code_size.py.

slavapestov committed:
Reflection: Decode imported Objective-C classes and protocols as their own TypeRef

Right now we expect that every class and protocol has a field
descriptor that tells us if the entity is @objc or not.

For imported types, the descriptor will not exist if we did not
directly emit a field whose concrete type contains the imported
type. For example, in lldb, we might have a generic type whose
runtime substituted type includes an imported type.

In this case, TypeLowering would fail to produce a layout because
it did not find a field descriptor for the imported type.

A better approach is to have the TypeDecoder call a different
factory method for imported types, and handle them specially in
TypeLowering, bypassing the field type metadata altogether.

slavapestov committed:
IRGen: Don't emit field descriptors for imported classes and protocols

slavapestov committed:
Fix MetadataReader handling of @objc protocols in existentials

rintaro committed:
[Lexer] Micro optimization for getTokenAt()

lex() invokes lexImpl(). It doesn't need to lex next token.

rintaro committed:
[CodeCompletion] Don't parse postfix if the primary ends with CC token

For example, given:

  [.foo(), .bar()]

If user want to insert another element in between:

  [.foo(), <HERE> .bar()]

'.bar()' is probably not a part of the inserting element. Moreover, having
the suffix doesn't help type inference in any way.

slavapestov committed:
Runtime: Add a FIXME

slavapestov committed:
RemoteAST: Add experimental swift::remoteAST::getTypeForMangling() entry point

This will eventually replace swift::ide::getTypeFromMangledSymbolname().

spevans committed:
SR-8876: Update for CMake build for Foundation

- Copy ICU licence into final install.

levivic committed:
Fix Runtime.swift.gyb failur.

drodriguez committed:
[android] Push test binaries w/o modifying executable name. (#19960)

At least test_rth.swift is checking the name of the executable
during the test, so renaming every executable to __executable in
Android will never work.

Also, during the rth tool execution, all the results from before
and after are pushed for every test. Since Android copies the
passed files without relative paths, the library files will
overwrite each other, making the test fail.

Depends on #19949 (more or less)

Fidetro committed:
Use the &>> operator in HashTest.

Gets rid of some obvious inefficiencies in this benchmark.

nkcsgexi committed:
Revert "[Build System: CMake] make add_swift_library a wrapper to add_llvm_library"

Dante-Broggi committed:
Fix typo in comment "eit" to "emit"

milseman commented:
Please test with following PR:
apple/swift-lldb#998

@swift-ci please test OS X Platform

slavapestov committed:
Move RemoteAST's RemoteASTTypeBuilder to lib/AST/ASTDemangler.cpp

slavapestov committed:
AST: Rename RemoteASTTypeBuilder to ASTBuilder

milseman commented:
Please test with following PR:
https://github.com/apple/swift-lldb/pull/1022

@swift-ci please test OS X Platform

shahmishal committed:
Update README to include Debian 9.5 under community-hosted CI.

shahmishal committed:
Fix the architecture for Debian 9.5

beccadax committed:
[PrintAsObjC] Disable -Watimport-in-framework-header warning in generated headers (#20086)

This is a new warning in clang when you use @import in a framework header. That’s often a mistake, but it isn’t in our generated -Swift.h headers. If we’re building with a clang that supports this warning, we’ll now emit a pragma to disable it.

beccadax committed:
Implementation for SE-0228: Fix ExpressibleByStringInterpolation (#20214)

* [CodeCompletion] Restrict ancestor search to brace

This change allows ExprParentFinder to restrict certain searches for parents to just AST nodes within the nearest surrounding BraceStmt. In the string interpolation rework, BraceStmts can appear in new places in the AST; this keeps code completion from looking at irrelevant context.

NFC in this commit, but keeps code completion from crashing once TapExpr is introduced.

* Remove test relying on ExpressibleByStringInterpolation being deprecated

Since soon enough, it won’t be anymore.

* [AST] Introduce TapExpr

TapExpr allows a block of code to to be inserted between two expressions, accessing and potentially mutating the result of its subexpression before giving it to its parent expression. It’s roughly equivalent to this function:

  func _tap<T>(_ value: T, do body: (inout T) throws -> Void) rethrows -> T {
    var copy = value
    try body(&copy)
    return copy
  }

Except that it doesn’t use a closure, so no variables are captured and no call frame is (even notionally) added.

This commit does not include tests because nothing in it actually uses TapExpr yet. It will be used by string interpolation.

* SE-0228: Fix ExpressibleByStringInterpolation

This is the bulk of the implementation of the string interpolation rework. It includes a redesigned AST node, new parsing logic, new constraints and post-typechecking code generation, and new standard library types and members.

* [Sema] Rip out typeCheckExpressionShallow()

With new string interpolation in place, it is no longer used by anything in the compiler.

* [Sema] Diagnose invalid StringInterpolationProtocols

StringInterpolationProtocol informally requires conforming types to provide at least one method with the base name “appendInterpolation” with no (or a discardable) return value and visibility at least as broad as the conforming type’s. This change diagnoses an error when a conforming type does not have a method that meets those criteria.

* [Stdlib] Fix map(String.init) source break

Some users, including some in the source compatibility suite, accidentally used init(stringInterpolationSegment:) by writing code like `map(String.init)`. Now that these intializers have been removed, the remaining initializers often end up tying during overload resolution. This change adds several overloads of `String.init(describing:)` which will break these ties in cases where the compiler previously selected `String.init(stringInterpolationSegment:)`.

* [Sema] Make callWitness() take non-mutable arrays

It doesn’t actually need to mutate them.

* [Stdlib] Improve floating-point interpolation performance

This change avoids constructing a String when interpolating a Float, Double, or Float80. Instead, we write the characters to a fixed-size buffer and then append them directly to the string’s storage.

This seems to improve performance for all three types, but especially for Double and Float80, which cannot always fit into a small string when stringified.

* [NameLookup] Improve MemberLookupTable invalidation

In rare cases usually involving generated code, an overload added by an extension in the middle of a file would not be visible below it if the type had lazy members and the same base name had already been referenced above the extension. This change essentially dirties a type’s member lookup table whenever an extension is added to it, ensuring the entries in it will be updated.

This change also includes some debugging improvements for NameLookup.

* [SILOptimizer] XFAIL dead object removal failure

The DeadObjectRemoval pass in SILOptimizer does not currently remove reworked string interpolations as well as the old design because their effects cannot be described by @_effects(readonly). That causes a test failure on Linux. This change temporarily silences that test. The SILOptimizer issue has been filed as SR-9008.

* Confess string interpolation’s source stability sins

* [Parser] Parse empty interpolations

Previously, the parser had an odd asymmetry which caused the same function to accept foo(), but reject “\()”. This change fixes the issue.

Already tested by test/Parse/try.swift, which uses this construct in one of its throwing interpolation tests.

* [Sema] Fix batch-mode-only lazy var bug

The temporary variable used by string interpolation needs to be recontextualized when it’s inserted into a synthesized getter. Fixes a compilation failure in Alamofire.

I’ll probably follow up on this bug a bit more after merging.

milseman committed:
[String] Assertion logic and isASCII bug fix.

Fix bugs in assertion logic and properly update the isASCII bit on
RRC. RRC tests added.

milseman committed:
[String] Drop in initial UTF-8 String prototype

This is a giant squashing of a lot of individual changes prototyping a
switch of String in Swift 5 to be natively encoded as UTF-8. It
includes what's necessary for a functional prototype, dropping some
history, but still leaves plenty of history available for future
commits.

My apologies to anyone trying to do code archeology between this
commit and the one prior. This was the lesser of evils.

milseman committed:
[test] Disable effects-analysis String interpolation test.

Despite trying out variations in @_effects and @inlinable, the effects
analysis that shows that an unused String interpolation can be
whole-scale discarded is failing.

rdar://problem/45788348

milseman commented:
I'm not sure why GitHub isn't picking up my push I did yesterday...

milseman commented:
I'll try closing and re-opening, in case we hit some kind of maximal PR commentary length...
